### PR TITLE
Framework For Database Upgrades [v2]

### LINKED ISSUE #369: Carrying Over Favorites From 1.x
The web version of PVTrAck does this without issue.

The mobile version, however, is another story.  The way we stored favorites/etc in 1.x was restricted to using the webview's implementation of localstorage.

In Ionic 2 the docs tell us this is no longer the case for the following reason:
> When running in a native app context, Storage will prioritize using SQLite, as it's one of the most stable and widely used file-based databases, and avoids some of the pitfalls of things like localstorage and IndexedDB, such as the OS deciding to clear out such data in low disk-space situations.

We can override this decision by telling the framework it **must** use localstorage and therefore enable our upgrading users to keep their favorites, but we run into the pitfalls described above.  This is a 60/40 tradeoff, really.  Either our users keep their favorites now and run the risk of losing them in the future, or they definitely lose them now but are significantly less likely to lose them in the future.

I'm interested in thoughts from @werebus, @sherson , and @petermarathas.  I won't ship 2.0 to the Play Store until a consensus is reached.

### PR DESCRIPTION
**Pull request by akaplo**
Closes #369.  Adds a file that manages database changes.  Also fixes a bug where saved trips from 1.x were causing crashes when being loaded. Thanks to @petermarathas for catching that.

### CODE REVIEW

File: src/app/storage.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import { Storage } from '@ionic/storage';
+import localforage from 'localforage';
+
+// INCREMENT THIS whenever you alter the database
+let currentVersion = 0;
```
Comment (werebus) [inline src/app/storage.ts]: I'm a bit confused by the version numbering here. Shouldn't `currentVersion` be something greater than 0?  That is, before we start, the version is `0`, and when we end the version is `0`

File: src/app/storage.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        // This would indicate that a user from 1.x is using 2.x
+        // for the first time.
+        if (returningUser) {
+          // Run all the migrations!
+          previousVersion = 0;
+        }
+        // For every database change added to the app since the user
+        // was last here, make those changes to their device's storage.
+        // New users will never enter the loop.
+        for (let version = previousVersion; version <= currentVersion; version++) {
```
Comment (werebus) [inline src/app/storage.ts]: So, likewise, here.  Should that be `<`

File: src/app/storage.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        localforage.iterate((value, key, iterationNumber) => {
+          console.log([key, value]);
+          this.storage.set(key, value);
+        }).then(() =>{
+          // In success, we no longer need 1.x's storage.
+          console.log('PVTrAck 1.x storage copy completed');
+          localforage.clear();
+          // ALWAYS update the user's current DB version after running
+          // a migration.
+          this.storage.set('previousVersion', currentVersion);
```
Comment (werebus) [inline src/app/storage.ts]: It also seems like you should set the the `previousVersion` to whatever migration we just ran.  As is, I think there are cases where we might skip over un-run migrations.

Comment (akaplo) [reply src/app/storage.ts]: Yeah, putting this line here was absolutely wrong.  How about setting `previousVersion` at the end of the loop (after it finishes)?

File: src/app/storage.ts (hunk 1 above)
Comment (akaplo) [reply src/app/storage.ts]: Fair enough. Changed `currentVersion` to 1.

**Review by petermarathas — APPROVED** [review]

**Comment by akaplo** [comment]
@werebus any final thoughts?

**Pull request merged**